### PR TITLE
replication: add rpc to get volumereplication info

### PR DESCRIPTION
**Pull request by yati1998**
This commit adds new rpc to get the volume
repliction info which at this moment, is mainly to
get the last_sync_time. we can update this later
as per the requirement.

Signed-off-by: yati1998 <ypadia@redhat.com>

### CODE REVIEW

**Comment by yati1998** [comment]
@ShyamsundarR @nixpanic @Rakshith-R please add your reviews here.

File: replication/README.md [hunk 1]
```diff
@@ -312,3 +315,37 @@ message ResyncVolumeResponse{
...
+  // this specific volume vs all other volumes supported by this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this volume.
+  string volume_id = 1;
+  // The identifier for the replication.
+  // This field is OPTIONAL.
+  // This field MUST contain enough information, together with volume_id,
+  // to uniquely identify this specific replication
+  // vs all other replications supported by this plugin.
+  string replication_id = 2 [(alpha_field) = true];
```
Comment (nixpanic) [inline replication/README.md:337]: why is this an _alpha_ field?

File: replication/README.md [hunk 2]
```diff
@@ -312,3 +315,37 @@ message ResyncVolumeResponse{
...
+  // Secrets required by the plugin to complete the request.
+  map<string, string> secrets = 4 [(csi.v1.csi_secret) = true];
+}
+
+// GetVolumeReplicationInfoResponse holds the information to send the
+// volume replication info.
+message GetVolumeReplicationInfoResponse{
+  // Holds the last sync time.
+  // This field is REQUIRED.
+  int64 last_sync_time = 1;
```
Comment (nixpanic) [inline replication/README.md:349]: what time formatting is this? that should be documented if there is no suitable time type.

Comment (nixpanic) [reply replication/README.md:349]: Using the same format as in the `message Snapshot` of the [CSI CreateSnapshot procedure](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot) would probably be better suitable.

```
  // Timestamp when the point-in-time snapshot is taken on the storage
  // system. This field is REQUIRED.
  .google.protobuf.Timestamp creation_time = 4;
```

You will need to import the proto file, where the other import statements are done:
```
import "google/protobuf/timestamp.proto";
```

File: replication/README.md [hunk 3]
```diff
@@ -312,3 +315,37 @@ message ResyncVolumeResponse{
...
+  string replication_id = 2 [(alpha_field) = true];
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  map<string, string> parameters = 3;
+  // Secrets required by the plugin to complete the request.
+  map<string, string> secrets = 4 [(csi.v1.csi_secret) = true];
+}
+
+// GetVolumeReplicationInfoResponse holds the information to send the
+// volume replication info.
+message GetVolumeReplicationInfoResponse{
```
Comment (nixpanic) [inline replication/README.md:346]: nit: formatting enhancement, space  before `{`.

You may also use an empty line before a comment+attribute block.

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: replication/README.md (hunk 1 above)
Comment (yati1998) [reply replication/README.md:337]: I just added similar to other rpcs, no specific reason

**Comment by yati1998** [comment]
@Madhu-1

File: replication/README.md (hunk 1 above)
Comment (nixpanic) [reply replication/README.md:337]: if it needs to be _alpha_, move it to the last field in the message

File: replication/README.md [hunk 4]
```diff
@@ -312,3 +316,38 @@ message ResyncVolumeResponse{
...
+  // this specific volume vs all other volumes supported by this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this volume.
+  string volume_id = 1;
+  // The identifier for the replication.
+  // This field is OPTIONAL.
+  // This field MUST contain enough information, together with volume_id,
+  // to uniquely identify this specific replication
+  // vs all other replications supported by this plugin.
+  string replication_id = 4 [(alpha_field) = true];
```
Comment (nixpanic) [inline replication/README.md:339]: please also move this to the end of the message. Newly added attributes (usually _alpha_ ones) should be appended, and not inserted at a random place.

**Review by nixpanic — DISMISSED** [review]

File: replication/README.md [hunk 5]
```diff
@@ -82,6 +83,9 @@ service Controller {
   // ResyncVolume RPC call to resync the volume.
   rpc ResyncVolume (ResyncVolumeRequest)
   returns (ResyncVolumeResponse) {}
+  // GetVolumeReplicationInfo RPC call to get the volume replication info.
```
Comment (Madhu-1) [inline replication/README.md:86]:
```suggestion
  // GetVolumeReplicationInfo RPC call to get the volume replication information.
```

File: replication/README.md [hunk 6]
```diff
@@ -312,3 +316,38 @@ message ResyncVolumeResponse{
...
 | Error is Unknown                                 | 2 UNKNOWN             | Indicates that a unknown error is generated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Caller MUST study the logs before retrying                                                                                                                                                                                            |
+
+### GetVolumeReplicationInfo
+
+```protobuf
+// GetVolumeReplicationInfoRequest holds the required information to get
+// the Volume replication info.
```
Comment (Madhu-1) [inline replication/README.md:324]:
```suggestion
// GetVolumeReplicationInfoRequest holds the required information to get
// the Volume replication information.
```

File: replication/README.md [hunk 7]
```diff
@@ -312,3 +316,38 @@ message ResyncVolumeResponse{
...
+  // The identifier for this volume, generated by the plugin during
+  // CreateVolume CSI RPC call.
+  // This field is REQUIRED.
+  // This field MUST contain enough information to uniquely identify
+  // this specific volume vs all other volumes supported by this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this volume.
+  string volume_id = 1;
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  map<string, string> parameters = 2;
```
Comment (Madhu-1) [inline replication/README.md:335]: why do we need parameters for Get call?

Comment (yati1998) [reply replication/README.md:335]: Not sure, just added a basic template based on other calls, if not required can update later?

Comment (Madhu-1) [reply replication/README.md:335]: Removing later can create inconsistency in the RPC pack and unpack. I would suggest removing it and adding it when required.

**Review by Madhu-1 — APPROVED** [review]

File: replication/README.md [hunk 8]
```diff
@@ -312,3 +317,36 @@ message ResyncVolumeResponse{
...
+  // The identifier for this volume, generated by the plugin during
+  // CreateVolume CSI RPC call.
+  // This field is REQUIRED.
+  // This field MUST contain enough information to uniquely identify
+  // this specific volume vs all other volumes supported by this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this volume.
+  string volume_id = 1;
+  // Secrets required by the plugin to complete the request.
+  map<string, string> secrets = 2 [(csi.v1.csi_secret) = true];
```
Comment (nixpanic) [inline replication/README.md:336]: As other calls like `ResyncVolumeRequest` contain the `parameters` map, it seems appropriate to pass there here as well. I imagine the parameters can contain things like hostnames of storage systems that are needed to obtain the volume replication information?

What is the reason to include it in `ResyncVolumeRequest`, but not here?

Comment (yati1998) [reply replication/README.md:336]: Not sure about it, but didn't find any specific use of it.  Maybe if @Madhu-1 can answer?

Comment (Madhu-1) [reply replication/README.md:336]: For actual set operations, you need parameters to set it on the resource, during get operation, we don't need any parameters. a volume_id should be enough to get the resource details.

Comment (nixpanic) [reply replication/README.md:336]: I do not understand what the difference between set/get are related to connecting to the storage provider. Even set-operations do not include parameters for the operation, only for the connection?

Could you explain with an example what parameters Promote/Demote/Resync might require, that are not needed for `GetVolumeReplicationInfo`? From my understanding, all these procedures are executed pretty much the same...

Comment (Madhu-1) [reply replication/README.md:336]: from the SPEC point of view, if you have the volumeID and the secret, you should be able to connect to the storage provider and do whatever option you want on the volume.
for Promote/Demote/Resync, we can pass any other storage provider-specific parameters for this function, the storage provide can take any action on it example we can have a force option where the storage provider can do some force promote/force demote etc.
For the Get option where do you think a parameter is required?

Comment (nixpanic) [reply replication/README.md:336]: Ok, I guess that makes sense. Thanks

**Review by nixpanic — APPROVED** [review]

**Pull request merged**